### DISCUSSION TITLE
requestLEScan(): do we have control over how the devices list is displayed?

### DISCUSSION THREAD

**Discussion by bonatoc** [capacitor-community/bluetooth-le#703] [Q&A] [answered]
Hello,

First of all thank you for this well-made library. 
We're about to publish a free app (iOS and Android) built upon it (Quasar 2 with Capacitor).

I was wondering if it was possible to alter/modify the scan result that appears in the OS pop-up.

The app scans the devices using `requestLEScan() `with `services`, which contains the manufacturer's service ID, and all is well.

**In iOS**, 
the pop-up lists (correctly) the devices by their name:
![Capture d’écran 2024-10-24 à 16 29 03](https://github.com/user-attachments/assets/26ea0d7e-f869-4db1-a7c5-76c2b4e8bb66)

**In Android**, 
the devices list in the popup looks like this (below are the exact same devices listed in the iOS pop-up above):
![Capture d’écran 2024-10-24 à 16 28 50](https://github.com/user-attachments/assets/ac6fd67e-7ec1-40ec-a7fe-3db627c45a7e)

It looks like iOS and Android are not looking for the same Generic Access (0x1800) characteristic.
**iOS** returns the value of Generic Access **0x2A00**, 
whereas **Android** returns the value of Generic Access **0x2A01** (hence the "Unknown" string returned)

Here's two NRF Connect screenshots:
![Capture d’écran 2024-10-24 à 16 24 05](https://github.com/user-attachments/assets/3e87cb38-e4cc-4f30-a990-ad1755c86c9d)
![Capture d’écran 2024-10-24 à 16 24 22](https://github.com/user-attachments/assets/55bb53c7-7729-4b33-89c4-1006cc0bdc3f)

Do we have any control over what's displayed in the Android "scan result" pop-up?
At the very least we would like to get rid of the MAC address of the device.

Thank you.

**Comment by peitschie** [comment]
Hi @bonatoc 

To eliminate the "Unknown" display in this list, you need to look at adding a `localName` value to the advertising packet from the peripheral.

The 0x1800 service you are showing above is **only** available after connection, so would not impact the display in this list, which can only use data in the advertising payloads. I suspect that the peripheral firmware is correctly setting the GAP name (which iOS will use), and not specifying an advertised local name, which Android uses.

You can confirm this by looking at the advertising payloads in nrf Connect. You can see in this screenshot here that there is a `Complete Local Name` field: https://play-lh.googleusercontent.com/yWNXh7QqlKDkbRDQyQjHqRI6_dc50Ww2GplGKVYdjZy7AFooMMeeMSJaHMXoiJ8aC3c=w2560-h1440

![unnamed](https://github.com/user-attachments/assets/d26261e1-7bcc-4ba6-8911-f5d69b1ac69a)


If you want further customisation of the display dialog (e.g., to remove the MAC address), you may want to consider implementing your own list using [requestLEScan](https://github.com/capacitor-community/bluetooth-le?tab=readme-ov-file#requestlescan) and Quasar.

**Reply by bonatoc** [reply]
Hi @peitschie ,

Thank you for you quick answer.

It seems that `localName` is provided in the advertising packet, see screencap (Android NRF Connect):
![Capture d’écran 2024-10-25 à 10 32 47](https://github.com/user-attachments/assets/3a6ded4e-d243-4f6e-93b6-c3a367b7a9ec)

So I think in the meantime I'm going to focus on the customisation of the display dialog.
Did you mean by that, I can completely bypass the native OS popup and implement my own?

Can you please point me to which files/functions have to be modified?
Is it the function itself, in `plugin.cjs.js` (`@capacitor-community/bluetooth-le/dist`)?
...or is rather in `BluetoothLe.kt` ? 


Thank you!

**Reply by peitschie** [reply, marked as answer]
>  Did you mean by that, I can completely bypass the native OS popup and implement my own?

Exactly this. [requestLEScan](https://github.com/capacitor-community/bluetooth-le?tab=readme-ov-file#requestlescan) takes a callback, that you can wire into your javascript UI toolkit of choice. In this case, I presume you'd feed this into a Quasar dialog of some description. There should be no need to modify anything within this plugin itself, as you are going to be writing in your web tech, not native code.

**Reply by bonatoc** [reply]
@peitschie  Many thanks for your time and for a great Capacitor plugin.
